# 1. Perkenalan


=================================================

Project 1

Nama  : Adhi Rizqi Alfaqih


Automasi ekstraksi data produk kemeja menggunakan Web Scraping dan transformasi data (Data Preparation) untuk mendukung pengambilan keputusan pembelian berdasarkan parameter popularitas, harga, dan rating.

=================================================

# 2. Case

Halo, saya Adhi. Saya menghadapi kendala karena hanya memiliki satu kemeja, sementara minggu ini terdapat banyak acara yang mewajibkan penggunaan kemeja. Oleh karena itu, saya berniat membeli kemeja baru. Saya menggunakan teknik web scraping untuk mencari kemeja dengan kategori 'Paling Populer', harga di bawah Rp500.000, dan memiliki rating produk terbaik.

# 3. Langkah Penyelesaian

## 3.1 Install bs4 selenium dan Pandas

In [ ]:
# Instalasi library Selenium untuk otomatisasi browser
!pip install bs4 selenium


In [ ]:
# Instalasi Pandas untuk pengolahan data
!pip install pandas

## 3.2 Import Pandas, BeautifulSoup, WebDriver, Time

In [ ]:
# Mengimpor Pandas untuk manipulasi dataframe
import pandas as pd
# Menggunakan BeautifulSoup untuk parsing HTML
from bs4 import BeautifulSoup
# Menggunakan Selenium untuk otomatisasi browser
from selenium import webdriver
# Menggunakan Requests untuk melakukan HTTP request ke website target
import requests
# Menggunakan time untuk memberikan jeda waktu pada saat perpindahan halaman
import time

import re

In [ ]:
# Melakukan validasi akses situs. Jika respon bernilai 200 (OK), maka situs tersebut dapat diproses untuk scraping
requests.get('https://www.zalora.co.id/c/men/pakaian/sale/c-26?vc_applicable=true&categoryId=31&page=1')

<Response [200]>

In [ ]:
# Untuk membuka browser otomatisasi
driver =  webdriver.Chrome()


In [ ]:
# Untuk menutup browser otomatisasi
driver.close()

## 3.3 Implementasi Web Scraping pada Zalora.co.id

In [ ]:
# Menjalankan webdriver
driver =  webdriver.Chrome()

# Membuat variable data, memiliki 5 data (kolom)
nama_produk = []
harga_produk = []
nama_toko = []
rating_produk = []
trending = []

# Melakukan scrapping hingga halaman ke 10
for page in range(1,11):
    # Memasukkan link yang dituju
    web_link = f'https://www.zalora.co.id/c/men/pakaian/sale/c-26?vc_applicable=true&categoryId=31&page={page}'
    driver.get(web_link)
    # melakukan jeda 1s
    time.sleep(1)
    # browser membuka halaman yang dituju
    html = driver.page_source
    # variable untuk mengambil html
    soup = BeautifulSoup(html, 'html.parser')
    # variable yang berisikan mengambil semua data pada 1 kotak produk
    kotak_produk = soup.find_all('div', {'class': 'flex flex-col min-h-[358px] tablet:min-h-[380px] desktop:min-h-[500px]'})

# Memasukkan data ke variable kotak_produk
    for data in kotak_produk:
        try:
            # scrapping html 'Nama Produk' ke dalam data
            print(data.find('div', {'class' : "overflow-hidden text-ellipsis whitespace-nowrap text-sm"}))
            nama_produk.append(data.find('div', {'class' : "overflow-hidden text-ellipsis whitespace-nowrap text-sm"}).get_text())
        except:
            #apabila tidak ada input, dialihkan menjadi 'None'
            print(None)
            nama_produk.append(None)
        try:
            # scrapping html 'Harga Produk' ke dalam data
            print(data.find('div', {"data-test-id":"originalPrice"}))
            harga_produk.append(data.find('div', {"data-test-id":"originalPrice"}).get_text())
        except:
            #apabila tidak ada input, dialihkan menjadi 'None'
            print(None)
            harga_produk.append(None)
        try:
            # scrapping html 'Nama Toko' ke dalam data
            print(data.find('span', {'class' : "shrink-[10] truncate text-base font-bold"}))
            nama_toko.append(data.find('span', {'class' : "shrink-[10] truncate text-base font-bold"}).get_text())
        except:
            #apabila tidak ada input, dialihkan menjadi 'None'
            print(None)
            nama_toko.append(None)
        try:
            # scrapping html 'Rating Produk' ke dalam data
            print(data.find('span', {"class" : "text-xxs font-medium leading-3 text-grey-60"}))
            rating_produk.append(data.find('span', {"class" : "text-xxs font-medium leading-3 text-grey-60"}).get_text())
        except:
            #apabila tidak ada input, dialihkan menjadi 'None'
            print(None)
            rating_produk.append(None)
        try:
            # scrapping html 'Trending' ke dalam data
            print(data.find('span', {"class" : "text-[#34669F]"}))
            trending.append(data.find('span', {"class" : "text-[#34669F]"}).get_text())
        except:
            #apabila tidak ada input, dialihkan menjadi 'None'
            print(None)
            trending.append(None)

<div class="overflow-hidden text-ellipsis whitespace-nowrap text-sm" data-test-id="productTitle">Leopold Kemeja Printing Pria Tree Motive Lengan Pendek Material Rayon ORIGINAL - Multicolor</div>
<div class="text-grey-60 line-through" data-test-id="originalPrice">Rp 1.422.500</div>
<span class="shrink-[10] truncate text-base font-bold" data-test-id="productBrandName" title="Hamlin">Hamlin</span>
<span class="text-xxs font-medium leading-3 text-grey-60" data-test-id="rating">4.0</span>
None
None
<div class="overflow-hidden text-ellipsis whitespace-nowrap text-sm" data-test-id="productTitle">Cotton Well Light Denim Series in Washed Black | Kemeja Lengan Panjang Denim Hitam</div>
<div class="text-grey-60 line-through" data-test-id="originalPrice">Rp 329.000</div>
<span class="shrink-[10] truncate text-base font-bold" data-test-id="productBrandName" title="Cotton Well">Cotton Well</span>
<span class="text-xxs font-medium leading-3 text-grey-60" data-test-id="rating">5.0</span>
None
None
<di

## 3.4 Memasukkan Hasil Scrapping ke DataFrame

In [29]:
#membuat data frame, dan memasukkan data scrapping ke tiap kolom
df = pd.DataFrame()
df['Nama Produk'] = nama_produk
df['Harga Produk'] = harga_produk
df['Nama Toko'] = nama_toko
df['Rating Produk'] = rating_produk
df['Paling Populer'] = trending


In [30]:
#output awal
df

Nama Produk  Harga Produk  \
0    Leopold Kemeja Printing Pria Tree Motive Lenga...  Rp 1.422.500   
1    Cotton Well Light Denim Series in Washed Black...    Rp 329.000   
2                                    Drury Denim Shirt    Rp 599.000   
3                               Dillion 2 Oxford Shirt    Rp 559.000   
4                            Tr Adelard Regular Jacket    Rp 999.000   
..                                                 ...           ...   
375                        Regular Fit Teddy overshirt    Rp 699.900   
376                 Santorini Wide Fit Denim Overshirt  Rp 1.049.000   
377                                       Henley Shirt  Rp 2.899.000   
378                    Basic Short Sleeves Linen Shirt    Rp 719.000   
379                            Slim Fit Collared Shirt    Rp 809.000   

        Nama Toko Rating Produk  Paling Populer  
0          Hamlin           4.0            None  
1     Cotton Well           5.0            None  
2    Hush Puppies           5.0  Paling Populer  
3    Hush Puppies           4.8            None  
4    Hush Puppies           5.0  Paling Populer  
..            ...           ...             ...  
375           H&M           4.8  Paling Populer  
376  Jack & Jones          None            None  
377          BOSS          None  Paling Populer  
378       DIVERSE          None            None  
379      Trendyol           4.6            None  

[380 rows x 5 columns]

## 3.5 Mengolah DataFrame

Data mentah hasil scraping seringkali mengandung karakter yang tidak diperlukan (seperti simbol mata uang atau teks tambahan). Tahap ini bertujuan untuk membersihkan data agar siap dianalisis secara numerik.

In [31]:
# untuk menghindari value dari duplikat, maka lakukan hapus duplikat dengan syntax ini
df = df.drop_duplicates()

In [32]:
# output sementara 1
df



Nama Produk  Harga Produk  \
0    Leopold Kemeja Printing Pria Tree Motive Lenga...  Rp 1.422.500   
1    Cotton Well Light Denim Series in Washed Black...    Rp 329.000   
2                                    Drury Denim Shirt    Rp 599.000   
3                               Dillion 2 Oxford Shirt    Rp 559.000   
4                            Tr Adelard Regular Jacket    Rp 999.000   
..                                                 ...           ...   
374                         EQ Zed solid RXD ss STREET    Rp 299.900   
376                 Santorini Wide Fit Denim Overshirt  Rp 1.049.000   
377                                       Henley Shirt  Rp 2.899.000   
378                    Basic Short Sleeves Linen Shirt    Rp 719.000   
379                            Slim Fit Collared Shirt    Rp 809.000   

        Nama Toko Rating Produk  Paling Populer  
0          Hamlin           4.0            None  
1     Cotton Well           5.0            None  
2    Hush Puppies           5.0  Paling Populer  
3    Hush Puppies           4.8            None  
4    Hush Puppies           5.0  Paling Populer  
..            ...           ...             ...  
374           H&M           4.9  Paling Populer  
376  Jack & Jones          None            None  
377          BOSS          None  Paling Populer  
378       DIVERSE          None            None  
379      Trendyol           4.6            None  

[321 rows x 5 columns]

In [ ]:
# terdapat perbedaan nilai baris pada output awal dan output yang sudah dihapus duplikatnya.
# menandakan bahwa menghapus duplikat telah berhasil.

In [33]:
# jika dilihat output, terdapat value 'none', menandakan terdapat missingvalues.
# saya akan mengubah missing values pada kolom 'Harga Produk' dari none menjadi '0'
df.loc[df['Harga Produk'].isna(), 'Harga Produk'] = "0"

In [34]:
# saya akan mengubah missing values pada kolom 'Rating Produk' dari none menjadi '0'
df.loc[df['Rating Produk'].isna(), 'Rating Produk'] = "0"

In [35]:
# saya akan mengubah missing values pada kolom 'Paling Populer' dari none menjadi 'Reguler'
df.loc[df['Paling Populer'].isna(), 'Paling Populer'] = "Reguler"


In [36]:
# untuk menghilangkan 'Rp.' dan '.'
df['Harga Produk'] = df['Harga Produk'].str.replace('Rp', '', regex = False).str.replace('.', '', regex = False)


C:\Users\User\AppData\Local\Temp\ipykernel_12036\1305173326.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Harga Produk'] = df['Harga Produk'].str.replace('Rp', '', regex = False).str.replace('.', '', regex = False)


In [37]:
# output sementara 2
df

Nama Produk Harga Produk  \
0    Leopold Kemeja Printing Pria Tree Motive Lenga...      1422500   
1    Cotton Well Light Denim Series in Washed Black...       329000   
2                                    Drury Denim Shirt       599000   
3                               Dillion 2 Oxford Shirt       559000   
4                            Tr Adelard Regular Jacket       999000   
..                                                 ...          ...   
374                         EQ Zed solid RXD ss STREET       299900   
376                 Santorini Wide Fit Denim Overshirt      1049000   
377                                       Henley Shirt      2899000   
378                    Basic Short Sleeves Linen Shirt       719000   
379                            Slim Fit Collared Shirt       809000   

        Nama Toko Rating Produk  Paling Populer  
0          Hamlin           4.0         Reguler  
1     Cotton Well           5.0         Reguler  
2    Hush Puppies           5.0  Paling Populer  
3    Hush Puppies           4.8         Reguler  
4    Hush Puppies           5.0  Paling Populer  
..            ...           ...             ...  
374           H&M           4.9  Paling Populer  
376  Jack & Jones             0         Reguler  
377          BOSS             0  Paling Populer  
378       DIVERSE             0         Reguler  
379      Trendyol           4.6         Reguler  

[321 rows x 5 columns]

In [38]:
# lalu kita akan mengecek tipe data
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 321 entries, 0 to 379
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Nama Produk     321 non-null    object
 1   Harga Produk    321 non-null    object
 2   Nama Toko       321 non-null    object
 3   Rating Produk   321 non-null    object
 4   Paling Populer  321 non-null    object
dtypes: object(5)
memory usage: 15.0+ KB


In [160]:
# dapat terlihat bahwa tipe data yaitu string (object) semua.
# untuk lebih mudah mengolah data, kolom yang memiliki angka, akan kita ubah menjadi int dan float


In [39]:
# merubah tipe data dari kolom 'Harga Produk' ke integer
df['Harga Produk'] = df['Harga Produk'].astype(int)

C:\Users\User\AppData\Local\Temp\ipykernel_12036\926880803.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Harga Produk'] = df['Harga Produk'].astype(int)


In [40]:
# merubah tipe data dari kolom 'Rating Produk' ke float
df['Rating Produk'] = df['Rating Produk'].astype(float)

C:\Users\User\AppData\Local\Temp\ipykernel_12036\259685042.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Rating Produk'] = df['Rating Produk'].astype(float)


In [41]:
# lalu kita cek kembali tipe data yang sudah kita rubah
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 321 entries, 0 to 379
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Nama Produk     321 non-null    object 
 1   Harga Produk    321 non-null    int32  
 2   Nama Toko       321 non-null    object 
 3   Rating Produk   321 non-null    float64
 4   Paling Populer  321 non-null    object 
dtypes: float64(1), int32(1), object(3)
memory usage: 13.8+ KB


In [ ]:
# untuk kolom "Harga Produk" dan 'Rating Produk' telah berhasil dirubah

In [48]:
# output sementara 3
df

Nama Produk  Harga Produk  \
0    Leopold Kemeja Printing Pria Tree Motive Lenga...       1422500   
1    Cotton Well Light Denim Series in Washed Black...        329000   
2                                    Drury Denim Shirt        599000   
3                               Dillion 2 Oxford Shirt        559000   
4                            Tr Adelard Regular Jacket        999000   
..                                                 ...           ...   
374                         EQ Zed solid RXD ss STREET        299900   
376                 Santorini Wide Fit Denim Overshirt       1049000   
377                                       Henley Shirt       2899000   
378                    Basic Short Sleeves Linen Shirt        719000   
379                            Slim Fit Collared Shirt        809000   

        Nama Toko  Rating Produk  Paling Populer  
0          Hamlin            4.0         Reguler  
1     Cotton Well            5.0         Reguler  
2    Hush Puppies            5.0  Paling Populer  
3    Hush Puppies            4.8         Reguler  
4    Hush Puppies            5.0  Paling Populer  
..            ...            ...             ...  
374           H&M            4.9  Paling Populer  
376  Jack & Jones            0.0         Reguler  
377          BOSS            0.0  Paling Populer  
378       DIVERSE            0.0         Reguler  
379      Trendyol            4.6         Reguler  

[321 rows x 5 columns]

In [ ]:
# Karna data terlalu banyak dan tidak mungkin akan dibeli semua, maka akan kita filterisasi

In [49]:
# Filterisasi #1 = dengan menampilkan hanya value "Paling Populer"
Paling_Populer_df = df[df["Paling Populer"] == "Paling Populer"]

In [50]:
# Filterisasi #2 = filter Harga Produk untuk menampilkan value kurang dari Rp.500.000
Harga_Produk_df = Paling_Populer_df[(Paling_Populer_df["Harga Produk"] < 500000) & (Paling_Populer_df["Harga Produk"] > 0)]

In [51]:
# Filterisasi #3 = menampilkan secara berurutan "Harga Produk" dari yang terendah ke yang terbesar
Asc_Harga_Produk_df = Harga_Produk_df.sort_values(by="Harga Produk", ascending=True)

In [52]:
# Filterisasi #4 = menampilkan secara berurutan "Rating Produk" dari yang terbesar ke yang terendah
Rating_df = Asc_Harga_Produk_df.sort_values(by="Rating Produk", ascending=False)

In [54]:
# Filterisasi #5 = hanya menampilkan daftar dari 5 teratas
# output final
Rating_df

Nama Produk  Harga Produk Nama Toko  Rating Produk  \
374  EQ Zed solid RXD ss STREET        299900       H&M            4.9   
345    LUCAS REG LINEN LS SHIRT        429900       H&M            4.9   
296               ELLIOT OVZ LS        499900       H&M            4.8   

     Paling Populer  
374  Paling Populer  
345  Paling Populer  
296  Paling Populer

Berdasarkan hasil pemrosesan data, ditemukan 3 produk H&M yang memenuhi kriteria. Produk "EQ Zed solid RXD ss Street" menjadi rekomendasi utama karena memiliki rating tertinggi di rentang harga di bawah Rp500.000.

# 4. Save this file

## 4.1 Export dataframe ke file .csv

In [ ]:
# export dataframe sebelum diolah ke file .csv
df.to_csv('DataFrame Zalora (Before).csv')

In [56]:
# export dataframe sesudah diolah ke file .csv
Rating_df.to_csv('DataFrame Zalora (After).csv')

## 4.2 Export ke SQL

Untuk menjamin integritas dan skalabilitas data, hasil akhir disimpan ke dalam format database relasional (SQLite) agar mudah diintegrasikan dengan sistem lain di masa mendatang

### 4.2.1 Install sqalchemy

In [57]:
!pip install sqlalchemy

### 4.2.2 Ekspor DataFrame ke Format Database (.db dan .sql)

In [61]:
from sqlalchemy import create_engine

# Membuat engine SQLite sementara bernama data.db
engine = create_engine("sqlite:///datamatang.db")

# Menyimpan DataFrame ke tabel bernama table_name di database SQLite.
Rating_df.to_sql("table_name", engine, if_exists="replace", index=False)

# Mengekspor database ke file .sql
with open("datamatang.sql", "w") as f:
    for line in engine.raw_connection().iterdump():
        f.write(f"{line}\n")


# 5. Kesimpulan

- Keberhasilan Ekstraksi Data: Program berhasil mengekstraksi data kemeja secara otomatis dengan parameter Nama Produk, Harga, Brand (Toko), Rating, dan Label Popularitas.

- Hasil Pencarian Optimal: Dari data yang dikumpulkan, ditemukan beberapa pilihan kemeja yang memenuhi kriteria utama, yaitu harga di bawah Rp500.000 dengan rating sempurna (5.0).

- Rekomendasi Utama:

    - Salah satu opsi terbaik adalah "EQ Zed solid RXD ss STREET" dengan harga Rp. 299.900 dan rating 4.9, yang sangat sesuai dengan anggaran dan ekspektasi kualitas.

    - Terdapat pula opsi lain seperti "LUCAS REG LINEN LS SHIRT" dan "ELLIOT OVZ LS"yang menawarkan harga kompetitif di bawah Rp. 500.000 dengan rating yang sangat baik.

- Pemanfaatan Data: Seluruh data yang telah dibersihkan kini tersimpan dalam format database relasional (datamatang.db dan datamatang.sql), sehingga memudahkan proses pemantauan harga atau integrasi data di masa mendatang.